File: aplv/web.py
"""Web UI サーバー."""

from __future__ import annotations

import json
import mimetypes
import re
import sqlite3
import threading
import urllib.parse
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path

from .discovery import find_log_files
from . import index
from .query import build_query_filter, parse_datetime, query_logs
from .path_util import normalize_path

STATIC_DIR = Path(__file__).parent / "static"


class LogViewerHandler(BaseHTTPRequestHandler):
    log_paths: list[Path] = []
    log_root: Path | None = None
    _conn: sqlite3.Connection | None = None
    load_status: str = "idle"
    load_error: str | None = None
    load_progress: int = 0
    _load_lock = threading.Lock()
    _db_lock = threading.Lock()

    def log_message(self, format: str, *args) -> None:  # noqa: A003
        return

    def _send_json(self, payload: object, status: int = 200) -> None:
        body = json.dumps(payload, ensure_ascii=False).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def _send_file(self, path: Path) -> None:
        if not path.exists():
            self.send_error(404)
            return
        content = path.read_bytes()
        mime, _ = mimetypes.guess_type(str(path))
        self.send_response(200)
        self.send_header("Content-Type", mime or "application/octet-stream")
        self.send_header("Content-Length", str(len(content)))
        self.end_headers()
        self.wfile.write(content)

    def _read_json_body(self) -> dict:
        length = int(self.headers.get("Content-Length", 0))
        if length <= 0:
            return {}
        raw = self.rfile.read(length)
        data = json.loads(raw.decode("utf-8"))
        if not isinstance(data, dict):
            raise ValueError("JSON オブジェクトを指定してください")
        return data

    @classmethod
    def _source_names(cls) -> list[str]:
        return [normalize_path(p) for p in cls.log_paths]

    @classmethod
    def _get_conn(cls) -> sqlite3.Connection:
        if cls._conn is None:
            cls._conn = index.open_memory()
        return cls._conn

    @classmethod
    def _replace_conn(cls, conn: sqlite3.Connection) -> None:
        with cls._db_lock:
            if cls._conn is not None and cls._conn is not conn:
                cls._conn.close()
            cls._conn = conn

    @classmethod
    def _start_load(cls) -> None:
        with cls._load_lock:
            if cls.load_status == "loading":
                return
            cls.load_status = "loading"
            cls.load_error = None
            cls.load_progress = 0

        root = cls.log_root
        paths = list(cls.log_paths)

        def worker() -> None:
            try:
                if root is None:
                    conn = index.open_memory()
                    index.clear_index(conn)
                    total = 0
                else:
                    index_store.ensure_tmp_dir_for(root)
                    conn = index.open_or_create(root)

                    def progress(count: int) -> None:
                        cls.load_progress = count

                    if not paths:
                        conn.close()
                        index_store.delete_index_files(root)
                        conn = index.open_or_create(root)
                        index.clear_index(conn)
                        total = 0
                    elif index.needs_rebuild(conn, paths):
                        conn.close()
                        index_store.delete_index_files(root)
                        conn = index.open_or_create(root)
                        total = index.build_index(conn, paths, progress)
                    else:
                        total = index.entry_count(conn)
                        cls.load_progress = total

                cls._replace_conn(conn)
                with cls._load_lock:
                    cls.load_status = "ready"
                    cls.load_progress = total
            except Exception as exc:  # noqa: BLE001
                with cls._load_lock:
                    cls.load_status = "error"
                    cls.load_error = str(exc)

        threading.Thread(target=worker, daemon=True).start()

    @classmethod
    def _ensure_load_started(cls) -> None:
        if cls.log_paths and cls.load_status == "idle":
            cls._start_load()

    @classmethod
    def _meta_payload(cls) -> dict:
        cls._ensure_load_started()
        loading = cls.load_status == "loading"
        progress = cls.load_progress

        if loading:
            total = progress
            first = None
            last = None
        elif cls.load_status == "ready":
            with cls._db_lock:
                conn = cls._get_conn()
                total = index.entry_count(conn)
                first, last = index.timestamp_bounds(conn)
        else:
            total = 0
            first = None
            last = None

        payload: dict = {
            "directory": str(cls.log_root) if cls.log_root else None,
            "files": cls._source_names(),
            "loading": loading,
            "load_status": cls.load_status,
            "load_progress": progress,
            "total": total,
            "first": first,
            "last": last,
        }
        if cls.load_error:
            payload["load_error"] = cls.load_error
        return payload

    def _browse_directory(self, raw_path: str) -> dict:
        if raw_path:
            current = Path(raw_path).expanduser().resolve()
        elif self.log_root is not None:
            current = self.log_root
        else:
            current = Path.cwd().resolve()

        if not current.is_dir():
            return {"error": f"ディレクトリが見つかりません: {current}"}

        parent = str(current.parent) if current.parent != current else None
        directories: list[str] = []
        try:
            for entry in sorted(current.iterdir(), key=lambda p: p.name.lower()):
                if entry.is_dir() and not entry.name.startswith("."):
                    directories.append(normalize_path(entry))
        except OSError as exc:
            return {"error": f"ディレクトリを読み取れません: {exc}"}

        return {
            "current": normalize_path(current),
            "parent": parent,
            "directories": directories,
        }

    def _load_directory(self, raw_path: str) -> tuple[dict, int]:
        if not raw_path:
            return {"error": "directory を指定してください"}, 400

        root = Path(raw_path).expanduser().resolve()
        if not root.is_dir():
            return {"error": f"ディレクトリが見つかりません: {root}"}, 400

        paths = find_log_files(root)
        LogViewerHandler.log_root = root
        LogViewerHandler.log_paths = paths
        LogViewerHandler.load_status = "idle"
        LogViewerHandler.load_error = None
        LogViewerHandler.load_progress = 0
        LogViewerHandler._start_load()
        return self._meta_payload(), 200

    def do_GET(self) -> None:  # noqa: N802
        parsed = urllib.parse.urlparse(self.path)
        params = urllib.parse.parse_qs(parsed.query)

        if parsed.path == "/":
            return self._send_file(STATIC_DIR / "index.html")
        if parsed.path.startswith("/static/"):
            rel = parsed.path.removeprefix("/static/")
            return self._send_file(STATIC_DIR / rel)

        if parsed.path == "/api/meta":
            return self._send_json(self._meta_payload())

        if parsed.path == "/api/browse":
            raw_path = params.get("path", [""])[0]
            payload = self._browse_directory(raw_path)
            if "error" in payload:
                return self._send_json(payload, 400)
            return self._send_json(payload)

        if parsed.path == "/api/logs":
            if self.load_status == "loading":
                return self._send_json(
                    {
                        "loading": True,
                        "load_progress": self.load_progress,
                        "total": 0,
                        "offset": 0,
                        "limit": 0,
                        "items": [],
                    }
                )
            if self.load_status == "error":
                return self._send_json(
                    {"error": self.load_error or "読み込みに失敗しました"},
                    500,
                )

            level = params.get("level", [""])[0]
            logger_pat = params.get("logger", [""])[0]
            thread_pat = params.get("thread", [""])[0]
            message_pat = params.get("message", [""])[0]
            grep = params.get("grep", [""])[0]
            source_pat = params.get("source", [""])[0]
            try:
                since_raw = params.get("since", [""])[0] or None
                until_raw = params.get("until", [""])[0] or None
                since = parse_datetime(since_raw) if since_raw else None
                until = parse_datetime(until_raw) if until_raw else None
            except ValueError as exc:
                return self._send_json({"error": str(exc)}, 400)
            try:
                limit = min(int(params.get("limit", ["500"])[0]), 5000)
                offset = max(int(params.get("offset", ["0"])[0]), 0)
            except ValueError:
                return self._send_json({"error": "limit/offset は整数で指定してください"}, 400)

            try:
                filt = build_query_filter(
                    level=level,
                    logger_pat=logger_pat,
                    thread_pat=thread_pat,
                    message_pat=message_pat,
                    grep=grep,
                    source_pat=source_pat,
                    since=since,
                    until=until,
                )
            except re.error as exc:
                return self._send_json({"error": str(exc)}, 400)

            with self._db_lock:
                total, page = query_logs(self._get_conn(), filt, offset, limit)

            return self._send_json(
                {
                    "total": total,
                    "offset": offset,
                    "limit": limit,
                    "items": [e.to_row_dict() for e in page],
                }
            )

        if parsed.path == "/api/logs/detail":
            source = params.get("source", [""])[0]
            timestamp = params.get("timestamp", [""])[0] or None
            try:
                line_no = int(params.get("line_no", ["0"])[0])
            except ValueError:
                return self._send_json({"error": "line_no は整数で指定してください"}, 400)
            if not source:
                return self._send_json({"error": "source を指定してください"}, 400)

            with self._db_lock:
                entry = index.find_entry(self._get_conn(), source, line_no, timestamp)
            if entry is None:
                return self._send_json({"error": "該当行が見つかりません"}, 404)

            try:
                raw = index.read_entry_raw(
                    Path(entry.source),
                    entry.byte_offset,
                    entry.end_byte_offset,
                )
            except OSError as exc:
                return self._send_json({"error": str(exc)}, 500)

            return self._send_json(
                {
                    "source": entry.source,
                    "line_no": entry.line_no,
                    "timestamp": entry.timestamp.isoformat(),
                    "level": entry.level,
                    "logger": entry.logger,
                    "thread": entry.thread,
                    "raw": raw,
                }
            )

        self.send_error(404)

    def do_POST(self) -> None:  # noqa: N802
        parsed = urllib.parse.urlparse(self.path)
        if parsed.path != "/api/load":
            self.send_error(404)
            return

        try:
            body = self._read_json_body()
        except (json.JSONDecodeError, UnicodeDecodeError, ValueError) as exc:
            return self._send_json({"error": str(exc)}, 400)

        payload, status = self._load_directory(body.get("directory", ""))
        return self._send_json(payload, status)


def serve(
    paths: list[Path],
    *,
    host: str = "127.0.0.1",
    port: int = 8766,
    log_root: Path | None = None,
) -> None:
    LogViewerHandler.log_paths = [p.resolve() for p in paths]
    LogViewerHandler.log_root = log_root.resolve() if log_root else None
    LogViewerHandler._conn = None
    LogViewerHandler.load_status = "idle"
    LogViewerHandler.load_error = None
    LogViewerHandler.load_progress = 0
    if paths:
        LogViewerHandler._start_load()
    server = ThreadingHTTPServer((host, port), LogViewerHandler)
    print(f"Application Log Viewer: http://{host}:{port}")
    if LogViewerHandler.log_root:
        print(f"ログディレクトリ: {LogViewerHandler.log_root}")
    print(f"読み込みファイル ({len(paths)}):")
    for p in paths:
        print(f"  - {p}")
    if not paths:
        print("  (未読み込み — ブラウザからディレクトリを選択してください)")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\n停止しました。")
    finally:
        server.server_close()
        with LogViewerHandler._db_lock:
            if LogViewerHandler._conn is not None:
                LogViewerHandler._conn.close()
                LogViewerHandler._conn = None

File: aplv-rs/src/web.rs
//! axum ベース Web サーバー（Python 版 API 互換）。
//!
//! - `GET /api/meta` — 読み込み状態
//! - `POST /api/load` — ディレクトリ指定・インデックス構築開始
//! - `GET /api/logs` — フィルタ付き一覧
//! - `GET /api/logs/detail` — スタックトレース含む生ログ
//!
//! インデックス構築は [`tokio::task::spawn_blocking`] でバックグラウンド実行。

use crate::discovery;
use crate::index;
use crate::path_util::normalize_path;
use crate::query::{self, QueryFilter};
use axum::body::Body;
use axum::extract::{Query, State};
use axum::http::{header, StatusCode};
use axum::response::{IntoResponse, Response};
use axum::routing::{get, post};
use axum::{Json, Router};
use serde::Deserialize;
use serde_json::{json, Value};
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::{Arc, Mutex};
use tokio::sync::RwLock;
use tower_http::services::ServeDir;

/// サーバー全体で共有する状態（ログパス・SQLite 接続・読み込み進捗）。
#[derive(Clone)]
pub struct AppState {
    pub log_root: Arc<RwLock<Option<PathBuf>>>,
    pub log_paths: Arc<RwLock<Vec<PathBuf>>>,
    pub load_status: Arc<Mutex<String>>,
    pub load_error: Arc<Mutex<Option<String>>>,
    pub load_progress: Arc<AtomicU64>,
    pub db: Arc<Mutex<rusqlite::Connection>>,
    pub static_dir: PathBuf,
}

impl AppState {
    pub fn new(static_dir: PathBuf) -> Self {
        Self {
            log_root: Arc::new(RwLock::new(None)),
            log_paths: Arc::new(RwLock::new(Vec::new())),
            load_status: Arc::new(Mutex::new("idle".to_string())),
            load_error: Arc::new(Mutex::new(None)),
            load_progress: Arc::new(AtomicU64::new(0)),
            db: Arc::new(Mutex::new(
                rusqlite::Connection::open_in_memory().expect("in-memory sqlite"),
            )),
            static_dir,
        }
    }

    fn status(&self) -> String {
        self.load_status.lock().expect("lock").clone()
    }

    pub async fn meta_payload(&self) -> Value {
        let log_root = self.log_root.read().await.clone();
        let files: Vec<String> = self
            .log_paths
            .read()
            .await
            .iter()
            .map(|p| normalize_path(p))
            .collect();
        let status = self.status();
        let loading = status == "loading";
        let progress = self.load_progress.load(Ordering::Relaxed);

        let (total, first, last) = if loading {
            (progress, None, None)
        } else if status == "ready" {
            let conn = self.db.lock().expect("db lock");
            let count = index::entry_count(&conn).unwrap_or(0);
            let bounds = index::timestamp_bounds(&conn).unwrap_or((None, None));
            (count, bounds.0, bounds.1)
        } else {
            (0, None, None)
        };

        let mut payload = json!({
            "directory": log_root.as_ref().map(|p| normalize_path(p)),
            "files": files,
            "loading": loading,
            "load_status": status,
            "load_progress": progress,
            "total": total,
            "first": first,
            "last": last,
        });
        if let Some(err) = self.load_error.lock().expect("lock").clone() {
            payload["load_error"] = json!(err);
        }
        payload
    }

    /// バックグラウンドスレッドでインデックス構築を開始する。
    pub async fn start_load(&self, root: PathBuf, paths: Vec<PathBuf>) {
        {
            let mut st = self.load_status.lock().expect("lock");
            if *st == "loading" {
                return;
            }
            *st = "loading".to_string();
        }
        *self.load_error.lock().expect("lock") = None;
        self.load_progress.store(0, Ordering::Relaxed);

        let state = self.clone();
        tokio::task::spawn_blocking(move || {
            if let Err(e) = state.load_blocking(root, paths) {
                *state.load_status.lock().expect("lock") = "error".to_string();
                *state.load_error.lock().expect("lock") = Some(e);
            }
        });
    }

    /// 同期的なインデックス構築（`spawn_blocking` 内から呼ぶ）。
    fn load_blocking(&self, root: PathBuf, paths: Vec<PathBuf>) -> Result<(), String> {
        index_store::ensure_tmp_dir_for(Some(&root));
        let mut conn = index::open_or_create(&root).map_err(|e| e.to_string())?;
        let rebuild = index::needs_rebuild(&conn, &paths).map_err(|e| e.to_string())?;
        let progress = self.load_progress.clone();
        let total = if paths.is_empty() {
            drop(conn);
            index_store::delete_index_files(&root);
            conn = index::open_or_create(&root).map_err(|e| e.to_string())?;
            index::clear_index(&conn).map_err(|e| e.to_string())?;
            0
        } else if rebuild {
            drop(conn);
            index_store::delete_index_files(&root);
            conn = index::open_or_create(&root).map_err(|e| e.to_string())?;
            index::build_index(&conn, &paths, |n| {
                progress.store(n, Ordering::Relaxed);
            })
            .map_err(|e| e.to_string())?
        } else {
            index::entry_count(&conn).map_err(|e| e.to_string())?
        };

        {
            let mut db = self.db.lock().map_err(|e| e.to_string())?;
            *db = conn;
        }

        self.load_progress.store(total, Ordering::Relaxed);
        *self.load_status.lock().map_err(|e| e.to_string())? = "ready".to_string();
        Ok(())
    }
}

pub fn router(state: AppState) -> Router {
    let static_dir = state.static_dir.clone();
    Router::new()
        .route("/", get(serve_index))
        .nest_service("/static", ServeDir::new(static_dir))
        .route("/api/meta", get(api_meta))
        .route("/api/browse", get(api_browse))
        .route("/api/load", post(api_load))
        .route("/api/logs", get(api_logs))
        .route("/api/logs/detail", get(api_logs_detail))
        .with_state(state)
}

async fn serve_index(State(state): State<AppState>) -> Result<Response, ApiError> {
    let path = state.static_dir.join("index.html");
    let body = tokio::fs::read(&path)
        .await
        .map_err(|e| ApiError::internal(format!("index.html: {e}")))?;
    Ok(Response::builder()
        .status(StatusCode::OK)
        .header(header::CONTENT_TYPE, "text/html; charset=utf-8")
        .body(Body::from(body))
        .expect("response"))
}

async fn api_meta(State(state): State<AppState>) -> Json<Value> {
    Json(state.meta_payload().await)
}

#[derive(Deserialize)]
struct BrowseQuery {
    path: Option<String>,
}

async fn api_browse(
    State(state): State<AppState>,
    Query(q): Query<BrowseQuery>,
) -> Result<Json<Value>, ApiError> {
    let current = if let Some(p) = q.path.filter(|s| !s.is_empty()) {
        PathBuf::from(&p)
    } else if let Some(root) = state.log_root.read().await.clone() {
        root
    } else {
        std::env::current_dir().map_err(|e| ApiError::bad(e.to_string()))?
    };
    Ok(Json(browse_directory(&current)?))
}

fn browse_directory(raw: &Path) -> Result<Value, ApiError> {
    let current = raw
        .canonicalize()
        .map_err(|_| ApiError::bad(format!("ディレクトリが見つかりません: {}", raw.display())))?;
    let current_display = normalize_path(&current);
    if !current.is_dir() {
        return Err(ApiError::bad(format!(
            "ディレクトリが見つかりません: {}",
            current_display
        )));
    }
    let parent = current
        .parent()
        .filter(|p| *p != current)
        .map(|p| normalize_path(p));
    let mut directories = Vec::new();
    let entries = std::fs::read_dir(&current).map_err(|e| ApiError::bad(format!("ディレクトリを読み取れません: {e}")))?;
    for entry in entries.flatten() {
        let path = entry.path();
        if path.is_dir() {
            let name = path.file_name().and_then(|n| n.to_str()).unwrap_or("");
            if !name.starts_with('.') {
                if let Ok(canon) = path.canonicalize() {
                    directories.push(normalize_path(&canon));
                }
            }
        }
    }
    directories.sort_by_key(|a| a.to_lowercase());
    Ok(json!({
        "current": current_display,
        "parent": parent,
        "directories": directories,
    }))
}

#[derive(Deserialize)]
struct LoadBody {
    directory: String,
}

async fn api_load(
    State(state): State<AppState>,
    Json(body): Json<LoadBody>,
) -> Result<(StatusCode, Json<Value>), ApiError> {
    if body.directory.is_empty() {
        return Err(ApiError::bad("directory を指定してください".to_string()));
    }
    let root = PathBuf::from(&body.directory)
        .canonicalize()
        .map_err(|_| ApiError::bad(format!("ディレクトリが見つかりません: {}", body.directory)))?;
    if !root.is_dir() {
        return Err(ApiError::bad(format!(
            "ディレクトリが見つかりません: {}",
            root.display()
        )));
    }
    let paths = discovery::find_log_files(&root).map_err(|e| ApiError::bad(e.to_string()))?;
    *state.log_root.write().await = Some(root.clone());
    *state.log_paths.write().await = paths.clone();
    *state.load_status.lock().expect("lock") = "idle".to_string();
    state.start_load(root, paths).await;
    Ok((StatusCode::OK, Json(state.meta_payload().await)))
}

#[derive(Deserialize)]
struct LogsQuery {
    level: Option<String>,
    logger: Option<String>,
    thread: Option<String>,
    message: Option<String>,
    grep: Option<String>,
    source: Option<String>,
    since: Option<String>,
    until: Option<String>,
    limit: Option<String>,
    offset: Option<String>,
}

async fn api_logs(
    State(state): State<AppState>,
    Query(q): Query<LogsQuery>,
) -> Result<Json<Value>, ApiError> {
    let status = state.status();
    if status == "loading" {
        let progress = state.load_progress.load(Ordering::Relaxed);
        return Ok(Json(json!({
            "loading": true,
            "load_progress": progress,
            "total": 0,
            "offset": 0,
            "limit": 0,
            "items": [],
        })));
    }
    if status == "error" {
        let err = state
            .load_error
            .lock()
            .expect("lock")
            .clone()
            .unwrap_or_else(|| "読み込みに失敗しました".to_string());
        return Err(ApiError::internal(err));
    }

    let limit = q
        .limit
        .as_deref()
        .unwrap_or("500")
        .parse::<u64>()
        .map(|n| n.min(5000))
        .map_err(|_| ApiError::bad("limit/offset は整数で指定してください".to_string()))?;
    let offset = q
        .offset
        .as_deref()
        .unwrap_or("0")
        .parse::<u64>()
        .map_err(|_| ApiError::bad("limit/offset は整数で指定してください".to_string()))?;

    let filter = build_filter(&q)?;
    let conn = state.db.lock().map_err(|e| ApiError::internal(e.to_string()))?;
    let (total, page) = query::query_logs(&conn, &filter, offset, limit)
        .map_err(|e| ApiError::internal(e.to_string()))?;
    let items: Vec<Value> = page.iter().map(query::entry_to_json).collect();
    Ok(Json(json!({
        "total": total,
        "offset": offset,
        "limit": limit,
        "items": items,
    })))
}

fn build_filter(q: &LogsQuery) -> Result<QueryFilter, ApiError> {
    let since = q
        .since
        .as_deref()
        .filter(|s| !s.is_empty())
        .map(query::parse_datetime)
        .transpose()
        .map_err(ApiError::bad)?;
    let until = q
        .until
        .as_deref()
        .filter(|s| !s.is_empty())
        .map(query::parse_datetime)
        .transpose()
        .map_err(ApiError::bad)?;

    Ok(QueryFilter {
        levels: q
            .level
            .as_deref()
            .filter(|s| !s.is_empty())
            .and_then(query::parse_level_filter),
        logger_re: opt_regex(q.logger.as_deref())?,
        thread_re: opt_regex(q.thread.as_deref())?,
        message_re: opt_regex(q.message.as_deref())?,
        source_re: opt_regex(q.source.as_deref())?,
        grep_re: opt_regex(q.grep.as_deref())?,
        since,
        until,
    })
}

fn opt_regex(pat: Option<&str>) -> Result<Option<regex::Regex>, ApiError> {
    match pat.filter(|s| !s.is_empty()) {
        Some(p) => query::compile_regex(p)
            .map(Some)
            .map_err(ApiError::bad),
        None => Ok(None),
    }
}

#[derive(Deserialize)]
struct DetailQuery {
    source: String,
    line_no: String,
    timestamp: Option<String>,
}

async fn api_logs_detail(
    State(state): State<AppState>,
    Query(q): Query<DetailQuery>,
) -> Result<Json<Value>, ApiError> {
    if q.source.is_empty() {
        return Err(ApiError::bad("source を指定してください".to_string()));
    }
    let line_no: i64 = q
        .line_no
        .parse()
        .map_err(|_| ApiError::bad("line_no は整数で指定してください".to_string()))?;
    let conn = state.db.lock().map_err(|e| ApiError::internal(e.to_string()))?;
    let entry = index::find_entry(
        &conn,
        &q.source,
        line_no,
        q.timestamp.as_deref().filter(|s| !s.is_empty()),
    )
    .map_err(|e| ApiError::internal(e.to_string()))?
    .ok_or_else(|| ApiError::not_found("該当行が見つかりません".to_string()))?;

    let raw = index::read_entry_raw(
        Path::new(&entry.source),
        entry.byte_offset as u64,
        entry.end_byte_offset.map(|v| v as u64),
    )
    .map_err(|e| ApiError::internal(e.to_string()))?;

    Ok(Json(json!({
        "source": entry.source,
        "line_no": entry.line_no,
        "timestamp": entry.timestamp.format("%Y-%m-%dT%H:%M:%S%.6f").to_string(),
        "level": entry.level,
        "logger": entry.logger,
        "thread": entry.thread,
        "raw": raw,
    })))
}

struct ApiError {
    status: StatusCode,
    message: String,
}

impl ApiError {
    fn bad(msg: String) -> Self {
        Self {
            status: StatusCode::BAD_REQUEST,
            message: msg,
        }
    }
    fn not_found(msg: String) -> Self {
        Self {
            status: StatusCode::NOT_FOUND,
            message: msg,
        }
    }
    fn internal(msg: String) -> Self {
        Self {
            status: StatusCode::INTERNAL_SERVER_ERROR,
            message: msg,
        }
    }
}

impl IntoResponse for ApiError {
    fn into_response(self) -> Response {
        (self.status, Json(json!({ "error": self.message }))).into_response()
    }
}
